test/test_typechecker.cpp

```diff
@@ -0,0 +1,97 @@
+#include "utap/typechecker.h"
+#include "utap/utap.h"
+
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
+#include <doctest/doctest.h>
+
+static constexpr const char* system_template = R""""(<?xml version="1.0" encoding="utf-8"?>
+<!DOCTYPE nta PUBLIC '-//Uppaal Team//DTD Flat System 1.5//EN' 'http://www.it.uu.se/research/group/darts/uppaal/flat-1_5.dtd'>
+<nta>
+	<declaration>%s</declaration>
+	<template>
+		<name x="5" y="5">Template</name>
+		<declaration>%s</declaration>
+		<location id="id0" x="0" y="0">
+		</location>
+		<init ref="id0"/>
+	</template>
+	<system>
+%s
+Process = Template();
+system Process;
+    </system>
+	<queries>
+		<query>
+			<formula/>
+			<comment/>
+		</query>
+	</queries>
+</nta>
+)"""";
+
+class document_fixture
+{
+    std::string decls{};
+    std::string template_decls{};
+    std::string sys_decls{};
+
+    template <typename... Args>
+    std::string string_format(const std::string& format, Args... args)
+    {
+        int size_s = std::snprintf(nullptr, 0, format.c_str(), args...) + 1;  // Extra space for '\0'
+        if (size_s <= 0) {
+            throw std::runtime_error("Error during formatting.");
+        }
+        auto size = static_cast<size_t>(size_s);
+        std::unique_ptr<char[]> buf(new char[size]);
+        std::snprintf(buf.get(), size, format.c_str(), args...);
```

Comment:
Better use std::string all the way:
```suggestion
        auto res = std::string(size, ' ');
        std::snprintf((char*)res.c_str(), size, format.c_str(), args...);
```
Note that `size+1` is no longer needed in this case